File: data_structures/python/QueueXX.py
class node:
    def __init__(self,data):
        self.data = data
        self.next = None
    
class QueueXX:
    def __init__(self):
        self.head = None

    def Enqueue(self,data):#insert last
        new_node = node(data)
        temp = self.head

        if(self.head == None):
            self.head = new_node
            return
        
        while(temp.next != None):
            temp = temp.next
        temp.next = new_node

    def Dequeue(self):

        if(self.head == None):
            print("Unable to Dequeue as queue is empty.\n")
            return

        temp = self.head
        iValue = temp

        self.head = self.head.next
        return iValue.data
    
    def Display(self):
        temp = self.head

        if(self.head == None):
            print("Unable to display as queue is empty.\n")
            return
        while(temp != None):
            print(f"| {temp.data} |",end = " ")
            temp = temp.next
        print("\n")

    def Count(self):
        temp = self.head
        iCount = 0

        if(self.head == None):
            print("Unable to Count as queue is empty.\n")
            return
        while(temp != None):
            iCount =  iCount + 1
            temp = temp.next
        return iCount
def main():
    sobj = QueueXX()

    sobj.Enqueue(10)
    sobj.Enqueue(20)
    sobj.Enqueue(30)
    sobj.Enqueue(40)
    sobj.Enqueue(50)

    sobj.Display()

    iRet = sobj.Count()

    print("The number of elements in the queue are: ",iRet)
    print("\n")

    iRet = sobj.Dequeue()
    print("The dequeuing element: ",iRet)

    sobj.Display()

    iRet = sobj.Dequeue()
    print("The dequeuing element: ",iRet)

    sobj.Display()


    iRet = sobj.Dequeue()
    print("The dequeuing element: ",iRet)

    sobj.Display()

    iRet = sobj.Count()

    print("The number of elements in the queue are: ",iRet)

if __name__ == "__main__":
    main()


        

File: data_structures/C++/QueueXX.cpp
#include<iostream>
using namespace std;

struct node
{
    int data;
    struct node * next;
};

typedef struct node NODE;
typedef struct node * PNODE;
//typedef struct node ** PPNODE;

class QueueXX
{
    private:
    PNODE first;
    int iCount;

    public:
    QueueXX()
    {
        cout<<"Inside constructor of QueueXX\n";
        this->first = NULL;
        this->iCount = 0;
    }

    void Enqueue(int no)//Insert Last
    { 
        PNODE newn = NULL;
        newn = new NODE;

        newn->data = no;
        newn->next = NULL;

        PNODE temp = NULL;

        if(first == NULL)
        {
            cout<<"Queue is empty\n";
            first = newn;
        }
        else
        {
            temp = first;
            while(temp->next != NULL)
            {
                temp = temp->next;
            }
            temp->next = newn;
        }
        iCount++;
    
    }

    int Dequeue()//Delete First
    {
        PNODE temp = NULL;
        int iValue = 0;

        if(first == NULL)//queue is empty
        {
            cout<<"Unable to dequeue as queue is empty.\n";
            return -1;
        }
        
          
        temp = first;
        iValue = temp->data;

        first = first->next;
        delete temp;
            
        iCount--;

        return iValue;
    
           
    }
    void Display()
    {
        PNODE temp = NULL;

        if(first == NULL)
        {
            cout<<"Unable to display elements as queue is empty\n";
            return;
        }
        
        
        temp = first;
        while(temp != NULL)
        {
            cout<<"| "<<temp->data<< " |";
            temp = temp->next;
        }
        cout<<"\n";
    }
    
    int Count()
    {
        return iCount;
    }
    
};

int main()
{
    QueueXX sobj;
    int iRet = 0;

    sobj.Enqueue(10);
    sobj.Enqueue(20);
    sobj.Enqueue(30);
    sobj.Enqueue(40);
    sobj.Enqueue(50);

    sobj.Display();

    iRet = sobj.Count();

    cout<<"Total number of elements in the queue are: "<<iRet<<"\n";

    iRet = sobj.Dequeue();
    cout<<"Dequeuing element : "<<iRet<<"\n";
    sobj.Display();

    iRet = sobj.Dequeue();
    cout<<"Dequeuing element : "<<iRet<<"\n";
    sobj.Display();

    iRet = sobj.Dequeue();
    cout<<"Dequeuing element : "<<iRet<<"\n";


    sobj.Display();

    iRet = sobj.Count();

    cout<<"Total number of elements in the queue are: "<<iRet<<"\n";
}

